### PR TITLE
fix(FindContractVersion): address web3.js inconsistant production behaviour with ethers.js

### PR DESCRIPTION
**Pull request by chrismaree**
**Motivation**

The most recent functionality with the perpetual bot support included the ability to automatically detect what contract version we are using. For some or other reason this failed to work in production.

**Summary**

After much painful testing, I was able to get this method to work in production by using ethers.js over web3.js. I'm not actually sure why this works, but it does 🤷 

**Details**

**Testing**

Check a box to describe how you tested these changes and list the steps for reviewers to test.

- [ ]  Ran end-to-end test, running the code as in production
- [ ]  New unit tests created
- [ ]  Existing tests adequate, no new tests required
- [x]  All existing tests pass
- [ ]  Untested

### CODE REVIEW

File: packages/common/src/ProviderUtils.js [hunk 1]
```diff
@@ -48,8 +48,8 @@ function createBasicProvider(url) {
  * @notice Gets a web3 instance based on the network argument using the truffle config in this package.
  * Use this for compatibility for running with or without truffle.
  * @example
- * // If a node app uses getWeb3() and you want to load network 1 with a default wallet
```
Comment (chrismaree) [inline packages/common/src/ProviderUtils.js:51]: this change is unrelated. the comment style was wrong.

File: packages/liquidator/index.js [hunk 2]
```diff
@@ -89,9 +89,11 @@ async function run({
       ).length == 0
     )
       throw new Error(
-        `Contract version specified or inferred is not supported by this bot. Provided/inferred config: ${JSON.stringify(
```
Comment (chrismaree) [inline packages/liquidator/index.js:92]: change made here helps with debugging as it prints out the whole `detectedContract`

File: packages/liquidator/test/Liquidator.js [hunk 3]
```diff
@@ -31,9 +31,9 @@ const { Liquidator } = require("../src/liquidator.js");
 // 2) non-matching 8 collateral & 18 synthetic for legacy UMA synthetics.
 // 3) matching 8 collateral & 8 synthetic for current UMA synthetics.
 const configs = [
-  { tokenSymbol: "WETH", collateralDecimals: 18, syntheticDecimals: 18, priceFeedDecimals: 18 }
-  // { tokenSymbol: "BTC", collateralDecimals: 8, syntheticDecimals: 18, priceFeedDecimals: 8 },
-  // { tokenSymbol: "BTC", collateralDecimals: 8, syntheticDecimals: 8, priceFeedDecimals: 18 }
+  { tokenSymbol: "WETH", collateralDecimals: 18, syntheticDecimals: 18, priceFeedDecimals: 18 },
```
Comment (chrismaree) [inline packages/liquidator/test/Liquidator.js:34]: un-related change. this should not have been commented out on master.

File: packages/liquidator/package.json [hunk 4]
```diff
@@ -32,7 +32,7 @@
   "bin": "index.js",
   "scripts": {
     "test": "yarn mocha-test && yarn hardhat-test",
-    "hardhat-test": "hardhat test --network hardhat",
```
Comment (chrismaree) [inline packages/liquidator/package.json:35]: this change was made as for some or other unknown reason having this `--network hardhat` broke the `ProviderUtils.js` with the following error:

image

we should consider removing this on all other packages as well but for now they seem to be passing CI.

File: packages/common/src/FindContractVersion.js [hunk 5]
```diff
@@ -3,16 +3,36 @@ const assert = require("assert");
...
   assert(web3, "Web3 object must be provided");
   assert(contractAddress, "Contract address must be provided");
-  const contractCode = await web3.eth.getCode(contractAddress);
+
+  // Note: there is an unknown issue in web3.js that means that the `getCode` syntax does not function correctly in
+  // production. However, ethers has proven to work correctly in production. The code below is a patch to still enable
+  // this module to work while we find a better long term solution for the web3.js issue. If running within unit tests
+  // then the web3.js version is required as it is scope according to the unit test.
+  let contractCode;
+  if (require("minimist")(process.argv.slice())?.network?.includes("mainnet")) {
```
Comment (mrice32) [inline packages/common/src/FindContractVersion.js:19]: nit: will this use web3 on testnets, like kovan? Will that break? Have you thought about using the isPublicNetwork function that we use in migrations?

A (potentially) more general way to switch is what we do in other parts of the codebase to determine whether we're in a truffle/hardhat test env:
```js
if (global.web3) {
  // This is run inside of truffle or hardhat test.
else {
  // This is run literally anywhere else.
}
```

**Review by mrice32 — COMMENTED** [review]
Looks awesome -- one minor question.

File: packages/common/src/FindContractVersion.js (hunk 5 above)
Comment (chrismaree) [reply packages/common/src/FindContractVersion.js:19]: ye, that's probably cleaner. The solution I had put here was specifically looking for the case where I knew this _did not_ work, which is mainnet. However, you are right that it is likely that this also does not work on kovan & there is no reason really to not generalize it further. 

Using your `if (global.web3) {` is probs the cleanest. I have updated it to your implementation, PTAL.

**Review by daywiss — APPROVED** [review]
Looks good

**Review by mrice32 — APPROVED** [review]
LGTM!
